src/main/java/com/microsoft/sqlserver/jdbc/SQLServerStatement.java

```diff
@@ -2410,79 +2423,98 @@ public final String getResponseBuffering() throws SQLServerException {
+            for (Map.Entry<String, SQLServerColumnEncryptionKeyStoreProvider> entry : clientKeyStoreProviders.entrySet()) {
+                String providerName = entry.getKey();
+                if (null == providerName || 0 == providerName.trim().length()) {
+                    throw new SQLServerException(null,
+                            SQLServerException.getErrString("R_EmptyCustomKeyStoreProviderName"), null, 0, false);
+                }
+
+                // MSSQL_CERTIFICATE_STORE not allowed on statement level
+                if ((providerName.equalsIgnoreCase(WINDOWS_KEY_STORE_NAME))) {
+                    MessageFormat form = new MessageFormat(
+                            SQLServerException.getErrString("R_InvalidCustomKeyStoreProviderName"));
+                    Object[] msgArgs = {providerName, WINDOWS_KEY_STORE_NAME};
+                    throw new SQLServerException(null, form.format(msgArgs), null, 0, false);
+                }
+
+                if (null == entry.getValue()) {
+                    throw new SQLServerException(null,
+                            String.format(SQLServerException.getErrString("R_CustomKeyStoreProviderValueNull"),
+                                    providerName), null, 0, false);
+                }
 
-            statementColumnEncryptionKeyStoreProviders.put(entry.getKey(), entry.getValue());
+                statementColumnEncryptionKeyStoreProviders.put(entry.getKey(), entry.getValue());
+            }
+        } finally {
+            lock.unlock();
         }
 
         loggerExternal.exiting(loggingClassName, "registerColumnEncryptionKeyStoreProvidersOnStatement",
                 "Number of statement-level Key store providers that are registered: "
                         + statementColumnEncryptionKeyStoreProviders.size());
     }
 
-    synchronized String getAllStatementColumnEncryptionKeyStoreProviders() {
-        String keyStores = "";
-        if (0 != statementColumnEncryptionKeyStoreProviders.size())
-            keyStores = statementColumnEncryptionKeyStoreProviders.keySet().toString();
-        return keyStores;
+    String getAllStatementColumnEncryptionKeyStoreProviders() {
+        lock.lock();
+        try {
+            String keyStores = "";
+            if (0 != statementColumnEncryptionKeyStoreProviders.size()) {
+                keyStores = statementColumnEncryptionKeyStoreProviders.keySet().toString();
+            }
+            return keyStores;
+        } finally {
+            lock.unlock();
+        }
     }
 
-    synchronized boolean hasColumnEncryptionKeyStoreProvidersRegistered() {
-        return null != statementColumnEncryptionKeyStoreProviders
-                && statementColumnEncryptionKeyStoreProviders.size() > 0;
+    boolean hasColumnEncryptionKeyStoreProvidersRegistered() {
+        lock.lock();
+        try {
+            return null != statementColumnEncryptionKeyStoreProviders && statementColumnEncryptionKeyStoreProviders.size() > 0;
```

Comment:
This doesn't exceed line length (80 chars)?

Reply:
I'll commit a change.